File: scripts/zigux/check-phase1-string-strlcat-anchors.py
#!/usr/bin/env python3
"""Guard the current Phase 1 string strlcat helper-local anchor packet against drift."""

from __future__ import annotations

import argparse
import tempfile
from pathlib import Path


HERE = Path(__file__).resolve()
DEFAULT_ROOT = HERE.parents[2] if len(HERE.parents) > 2 else HERE.parent
STRING_HELPER_REL = Path("tools/lib/string.zig")

REQUIRED_SYMBOLS = (
    "pub fn strlcat(dest: []u8, src: []const u8) usize {",
    "const src_len = cStringLen(src);",
    "const dest_len = strnlen(dest, dest.len);",
    "return dest.len + src_len;",
    "const copy_len = @min(src_len, dest.len - dest_len - 1);",
    "dest[dest_len + copy_len] = 0;",
    "return dest_len + src_len;",
)

REQUIRED_TEST_ANCHORS = (
    'test "strlcat appends within the destination size and reports the attempted length"',
    'test "strlcat truncates with a terminator and keeps the full attempted length"',
    'test "strlcat treats an unterminated destination as full"',
    'test "strlcat handles a zero-length destination buffer"',
)


def repo_root(root: str | None) -> Path:
    return Path(root).resolve() if root else DEFAULT_ROOT.resolve()


def helper_path(root: Path) -> Path:
    return root / STRING_HELPER_REL


def collect_failures(root: Path) -> list[str]:
    path = helper_path(root)
    if not path.exists():
        return [f"missing_file:{STRING_HELPER_REL.as_posix()}"]

    text = path.read_text(encoding="utf-8")
    failures: list[str] = []
    for symbol in REQUIRED_SYMBOLS:
        count = text.count(symbol)
        if count != 1:
            failures.append(f"string_strlcat_symbol:{symbol}:expected=1:actual={count}")
    for anchor in REQUIRED_TEST_ANCHORS:
        count = text.count(anchor)
        if count != 1:
            failures.append(f"string_strlcat_anchor:{anchor}:expected=1:actual={count}")
    return failures


def sample_helper_text() -> str:
    symbol_lines = "\n".join(REQUIRED_SYMBOLS)
    test_lines = "\n".join(REQUIRED_TEST_ANCHORS)
    return f"{symbol_lines}\n\n{test_lines}\n"


def write_sample_root(root: Path) -> None:
    path = helper_path(root)
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(sample_helper_text(), encoding="utf-8")


def run_self_test() -> int:
    case_count = 0

    with tempfile.TemporaryDirectory(prefix="phase1-string-strlcat-") as tmp:
        root = Path(tmp)

        failures = collect_failures(root)
        if failures != [f"missing_file:{STRING_HELPER_REL.as_posix()}"]:
            print("self-test:missing-file:unexpected-result")
            for item in failures:
                print(item)
            return 1
        case_count += 1

        write_sample_root(root)
        failures = collect_failures(root)
        if failures:
            print("self-test:baseline:unexpected-failures")
            for item in failures:
                print(item)
            return 1
        case_count += 1

        helper = helper_path(root)
        text = helper.read_text(encoding="utf-8")
        text = text.replace(REQUIRED_SYMBOLS[3] + "\n", "", 1)
        helper.write_text(text, encoding="utf-8")
        failures = collect_failures(root)
        expected = f"string_strlcat_symbol:{REQUIRED_SYMBOLS[3]}:expected=1:actual=0"
        if expected not in failures:
            print("self-test:missing-symbol:unexpected-result")
            for item in failures:
                print(item)
            return 1
        case_count += 1

        write_sample_root(root)
        text = helper.read_text(encoding="utf-8")
        duplicated = REQUIRED_TEST_ANCHORS[0]
        text = text.replace(duplicated + "\n", duplicated + "\n" + duplicated + "\n", 1)
        helper.write_text(text, encoding="utf-8")
        failures = collect_failures(root)
        expected = f"string_strlcat_anchor:{duplicated}:expected=1:actual=2"
        if expected not in failures:
            print("self-test:duplicate-anchor:unexpected-result")
            for item in failures:
                print(item)
            return 1
        case_count += 1

    print("PHASE1_STRING_STRLCAT_ANCHORS_SELF_TEST=pass")
    print(f"PHASE1_STRING_STRLCAT_ANCHORS_SELF_TEST_CASE_COUNT={case_count}")
    return 0


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--root", help="override the repository root used for validation")
    parser.add_argument(
        "--write-sample-root",
        help="write a minimal sample tree to this directory and exit",
    )
    parser.add_argument("--self-test", action="store_true", help="run checker self-tests")
    args = parser.parse_args()

    if args.write_sample_root:
        root = Path(args.write_sample_root).resolve()
        write_sample_root(root)
        print(f"PHASE1_STRING_STRLCAT_ANCHORS_SAMPLE_ROOT={root}")
        return 0

    if args.self_test:
        return run_self_test()

    failures = collect_failures(repo_root(args.root))
    if failures:
        for item in failures:
            print(item)
        return 1

    print("PHASE1_STRING_STRLCAT_ANCHORS=pass")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/zigux/check_phase1_string_strlcat_anchors.zig
// Ported from check-phase1-string-strlcat-anchors.py by port_phase1_guards.py
const std = @import("std");
const Io = std.Io;
const guard = @import("zigux_guard.zig");

pub const pass_marker = "PHASE1_STRING_STRLCAT_ANCHORS_SELF_TEST=pass";

const REQUIRED_SYMBOLS = [_][]const u8{
    "pub fn strlcat(dest: []u8, src: []const u8) usize {",
    "const src_len = cStringLen(src);",
    "const dest_len = strnlen(dest, dest.len);",
    "return dest.len + src_len;",
    "const copy_len = @min(src_len, dest.len - dest_len - 1);",
    "dest[dest_len + copy_len] = 0;",
    "return dest_len + src_len;",
};

const REQUIRED_TEST_ANCHORS = [_][]const u8{
    "test \"strlcat appends within the destination size and reports the attempted length\"",
    "test \"strlcat truncates with a terminator and keeps the full attempted length\"",
    "test \"strlcat treats an unterminated destination as full\"",
    "test \"strlcat handles a zero-length destination buffer\"",
};

const STRING_HELPER_REL = "tools/lib/string.zig";

fn collectFailures(
    io: Io,
    allocator: std.mem.Allocator,
    root: []const u8,
) !std.ArrayList([]const u8) {
    var failures: std.ArrayList([]const u8) = .empty;
    errdefer {
        for (failures.items) |item| allocator.free(item);
        failures.deinit(allocator);
    }
    {
        const relative_path = STRING_HELPER_REL;
        const full_path = try guard.joinPath(allocator, root, relative_path);
        defer allocator.free(full_path);
        if (!guard.pathExists(io, full_path)) {
            try guard.appendMissingFileIssue(allocator, &failures, relative_path);
            return failures;
        }
        const text = try guard.readUtf8File(io, allocator, full_path);
        defer allocator.free(text);
        for (REQUIRED_SYMBOLS) |symbol| {
            const count = guard.countOccurrences(text, symbol);
            if (count != 1) {
                const issue = try std.fmt.allocPrint(allocator, "string_strlcat_symbol:{s}:expected=1:actual={d}", .{ symbol, count });
                try failures.append(allocator, issue);
            }
        }
        for (REQUIRED_TEST_ANCHORS) |anchor| {
            const count = guard.countOccurrences(text, anchor);
            if (count != 1) {
                const issue = try std.fmt.allocPrint(allocator, "string_strlcat_anchor:{s}:expected=1:actual={d}", .{ anchor, count });
                try failures.append(allocator, issue);
            }
        }
    }

    return failures;
}

pub fn runSelfTest(io: Io, allocator: std.mem.Allocator) !u8 {
    var tmp = try guard.TempWorkspace.init(io, allocator, "selftest");
    defer tmp.deinit();
    const root = try tmp.rootPath(allocator);
    defer allocator.free(root);
    {
        const relative_path = "tools/lib/string.zig";
        const full_path = try guard.joinPath(allocator, root, relative_path);
        defer allocator.free(full_path);
        try guard.writeUtf8File(io, full_path, "\n");
    }
    var failures = try collectFailures(io, allocator, root);
    defer {
        for (failures.items) |item| allocator.free(item);
        failures.deinit(allocator);
    }
    try guard.expectSelfTest(failures.items.len == 0);
    try guard.printLine(io, "{s}", .{pass_marker});
    try guard.printLine(io, "PHASE1_STRING_STRLCAT_ANCHORS_SELF_TEST_CASE_COUNT={d}", .{@as(usize, 1)});
    return 0;
}

pub fn main(init: std.process.Init) !void {
    const allocator = init.gpa;
    const io = init.io;
    const args = try init.minimal.args.toSlice(allocator);

    var explicit_root: ?[]const u8 = null;
    var self_test = false;
    var index: usize = 1;
    while (index < args.len) : (index += 1) {
        const arg = args[index];
        if (std.mem.eql(u8, arg, "--self-test")) {
            self_test = true;
            continue;
        }
        if (std.mem.eql(u8, arg, "--root") or std.mem.eql(u8, arg, "--repo-root")) {
            if (index + 1 >= args.len) std.process.exit(2);
            index += 1;
            explicit_root = args[index];
            continue;
        }
    }

    if (self_test) {
        std.process.exit(try runSelfTest(io, allocator));
    }

    const root = if (explicit_root) |value| value else try guard.defaultRepoRoot(allocator);
    defer if (explicit_root == null) allocator.free(root);

    var failures = try collectFailures(io, allocator, root);
    defer {
        for (failures.items) |item| allocator.free(item);
        failures.deinit(allocator);
    }

    if (failures.items.len > 0) {
        try guard.printLine(io, "PHASE1_STRING_STRLCAT_ANCHORS_SELF_TEST=fail", .{});
        for (failures.items) |failure| try guard.printLine(io, "{s}", .{failure});
        std.process.exit(1);
    }

    try guard.printLine(io, "{s}", .{pass_marker});
    std.process.exit(0);
}
